### PR TITLE
Add experimental Result3 && SResult3 (Serializable Result)

### PR DESCRIPTION
**Pull request by sydhds**
* Result3 (similar to Result)
  * allow to build an Error variant without having to instantiate a default value (for T)
  * has staitc methods toOk(...) & toErr(...) that are very explicit
  * has protected members so we can derive from it (see SResult3)
* SResult3 is Result3<T> where T is Serializable:
  * implement Serializable & Deserializable  

+ add command: npm run test:types 
+ add unit testing for Result / Result3 / SResult3

This need careful review and a decision if we could rename Result3 to Result (or get rid of Result3)... :)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "test": "asp && npm run test:ts",
     "test:ci": "asp --summary --config as-pect-ci.config.js",
     "test:proba": "asp --summary --config as-pect-proba.config.js",
+    "test:types": "asp --summary --config as-pect-types.config.js",
```
Comment (peterjah) [inline package.json:27]: forgot to commit `as-pect-types.config.js`

Comment (sydhds) [reply package.json:27]: Done

**Review by Thykof — CHANGES_REQUESTED** [review]
I don't understand much the goal because I miss some context.

I thing there is something wrong in the unit test with `res2` and `sRes2`.

Overall, looks good

File: packages/as-types/assembly/__tests__/result.spec.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  });
+});
+
+describe('Result3 tests', () => {
+  it('Test 1: basics', () => {
+    // let d1 = new Divinity()
+    let d1 = new Divinity(14, 'Poseidon?!#');
+    let d2 = new Divinity(42, 'Zeus');
+    let res1: Result3<Divinity> = new Result3(d1);
+    let res2 = new Result3<Divinity>(new Divinity());
```
[the remark this answers is not in the document]
Comment (sydhds) [reply packages/as-types/assembly/__tests__/result.spec.ts]: fixed

File: packages/as-types/assembly/__tests__/result.spec.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    let res3 = Result3.toOk<Divinity>(d2);
+    expect<Divinity>(res1.unwrap()).toBe(d1);
+    expect<Divinity>(res3.unwrap()).toBe(d2);
+
+    let sRes1 = SResult3.fromResult(res1);
+    expect<Divinity>(sRes1.unwrap()).toBe(d1);
+    let sRes2 = SResult3.fromResult(res2);
+    expect<Divinity>(sRes1.unwrap()).toBe(d1);
```
[the remark this answers is not in the document]
Comment (sydhds) [reply packages/as-types/assembly/__tests__/result.spec.ts]: fixed

File: packages/as-types/assembly/__tests__/result.spec.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    let _Res1 = SResult3.toResult(sRes1);
+    expect<Divinity>(_Res1.unwrap()).toBe(d1);
+    expect<Divinity>(_Res1.expect('expect this & that')).toBe(d1);
```
[the remark this answers is not in the document]
Comment (sydhds) [reply packages/as-types/assembly/__tests__/result.spec.ts]: updated the error message

**Review by Thykof — DISMISSED** [review]

**Comment by gregLibert** [comment]
This issue is on hold. I'm removing myself as the assignee to declutter my to-do board.

**Comment by AurelienFT** [comment]
@Ben-Rey I have removed the serializable result class as it didn't work. It was using the fact that in the new `Result3` struct the `value` was nullable but it's impossible with `Result` being used with non nullable types such as u8, u16, ... The tests wasn't testing this.

Now that we come back to a version (like in main) where value is mandatory then the behavior of `nextSerializable()` isn't working with Result has it needs one parameter to instantiate the object. I tried to make a special case but the `add` function is hard to play with if we want to have nested template condition type such as `arg instance of Result<U: Serializable>`. I didn't found an elegant way to sort this out so I removed it. If it's important (or you have an idea) I can re-add it.

**Pull request closed without merging**